_episodes/31.md

```diff
@@ -0,0 +1,579 @@
+microservices in applications.
+
+The above example is showing a NodePort service. There are three service types.
+
+1. ClusterIP - the service creates a virtual ip inside the cluster to enable 
+communication between different services. This service is the default when you
+don't specify a `type` value under `spec` in the configuration.
+2. NodePort - is used to expose the internal address of a pod using the ip 
+address and port of the node it is running on.
+3. Load Balancer - this service creates a load balancer for the application in 
+supported cloud providers. We won't cover this one, but this is used when 
+setting launching in EKS using the `eksctl`.
+
+Here's a diagram of the ClusterIP networking between different ReplicaSets.
+<p align="center">
+ <img align="center" width="75%" height="100%" src="/assets/episode/31/clusterip.png"/><br/>
+ Source: https://www.udemy.com/course/learn-kubernetes/
+</p>
+
+While enabling a node port to establish connectivity to a specific ReplicaSet of 
+pod instances, it cannot make a generically accessible ip address for services 
+to communicate between one another.
+
+In our case, we want to make the coordinator node have an external ip address.
+The helm chart defines a ClusterIP service to accomplish this. Notice the
+selector targets the Trino app, the release label, and only the coordinator 
+component, which we know is one node.
+
+```
+apiVersion: v1
+kind: Service
+metadata:
+  name: tcb-trino
+  labels:
+    app: trino
+    chart: trino-0.3.0
+    release: tcb
+    heritage: Helm
+spec:
+  type: ClusterIP
+  ports:
+    - port: 8080
+      targetPort: http
+      protocol: TCP
+      name: http
+  selector:
+    app: trino
+    release: tcb
+    component: coordinator
+```
+
+### NodePort
+
+The [NodePort](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport) 
+Service type, creates a proxy service to forward traffic to a specific port on 
+the node from the pod.
+
+<p align="center">
+ <img align="center" width="75%" height="100%" src="/assets/episode/31/service.png"/><br/>
+ Source: https://www.udemy.com/course/learn-kubernetes/
+</p>
+
+There are three ports when setting up a NodePort. 
+
+1. TargetPort - is the port number on the pod itself, where the service forward to.
+2. Port - is the port used by the service. Which itself has an ip address
+3. NodePort - is the port that is exposed by the worker node and made available 
+externally. NodePorts can only be in the range of 30000 - 32767.
+
+The only required port to set is `port`. By default `targetPort` will be the 
+same as `port` and nodePort will get automatically assigned a free port in the 
+allowed range. `ports` is also an array which is why the `-` char is used. 
+
+
+### Amazon EKS (Elastic Kubernetes Service)
+
+Amazon EKS is a managed container service to run and scale Kubernetes 
+applications in the cloud or on-premises. EKS manages the control plane for you 
```

Comment:
remove "on-premises" imho

Reply:
I like this to drive the point home that workers are independent of the control plane and when you use EKS, or any of the other services, that is really the service being "cloudified".

Reply:
While that is theoretically true in the real world that just doesnt hold up unless you have very specific circumstances and knowledge. The service and the cluster have to be as much colocated as possible to have any reasonable performance. The whole point of using EKS is to make it simple .. with trying to hook in external machines on prem or wherever you are making it very complicated... its an edge case usage at best and I would leave it out to avoid confusion .

Reply:
So ... what are doing here? I dont think EKS is used much or at all with nodes outside of EKS . performance would be terrible. I would rephrase this section to say that EKS provides K8s clusters in the cloud for you without your having to manage the whole k8s services and platform .. which is a HUGE win. Running k8s yourself is really hard.